File: polyglot/python/vendor_registry.py
"""vendor_registry.py — Vendor Registry Module for vendorvet

A complete implementation of a third-party/vendor risk questionnaire system
with SBOM cross-referencing capabilities.

Usage:
    from vendor_registry import VendorRegistry, Vendor
    registry = VendorRegistry()
    registry.add_vendor(Vendor(...))
    results = registry.search(name="Acme Corp")
"""

from __future__ import annotations
import json
import csv
import hashlib
import logging
import os
from dataclasses import dataclass, field, asdict
from datetime import datetime, date
from enum import Enum, auto
from typing import (
    Optional, List, Dict, Any, Iterator, Tuple, 
    Callable, TypeVar, Generic, Protocol
)

# Configure module-level logging
_logger = logging.getLogger(__name__)


class RiskLevel(Enum):
    """Risk classification levels for vendors."""
    CRITICAL = auto()      # 90-100
    HIGH = auto()          # 75-89
    MEDIUM = auto()        # 60-74
    LOW = auto()           # 25-59
    MINIMAL = auto()       # 0-24


class ComplianceStatus(Enum):
    """Questionnaire compliance states."""
    PENDING = "pending"
    IN_PROGRESS = "in_progress"
    REVIEW = "review"
    APPROVED = "approved"
    CONDITIONAL = "conditional"
    FAILED = "failed"
    EXPIRED = "expired"


class SBOMFormat(Enum):
    """Supported SBOM standards."""
    SPDX = "spdx"
    CYCLONEDX = "cyclonedx"
    CPE = "cpe"
    CUSTOM = "custom"


@dataclass(frozen=True)
class Contact:
    """Vendor contact information."""
    name: str
    email: Optional[str] = None
    phone: Optional[str] = None
    website: Optional[str] = None
    
    def __post_init__(self):
        if not self.name.strip():
            raise ValueError("Contact name cannot be empty")


@dataclass(frozen=True)
class VendorProfile:
    """Core vendor profile data."""
    name: str
    industry: Optional[str] = None
    size_category: Optional[str] = None  # S/M/L/Enterprise
    headquarters: Optional[str] = None
    year_founded: Optional[int] = None
    
    @property
    def unique_id(self) -> str:
        """Generate a stable identifier from name."""
        return hashlib.sha256(
            self.name.strip().lower().encode()
        ).hexdigest()[:12]


@dataclass(frozen=True)
class SBOMReference:
    """Reference to an SBOM document."""
    vendor_name: str
    software_product: Optional[str] = None
    sbom_url: Optional[str] = None
    sbom_hash: Optional[str] = None  # SHA-256 of downloaded content
    format: SBOMFormat = SBOMFormat.CUSTOM
    version: Optional[str] = None
    last_verified: Optional[datetime] = None
    
    def __hash__(self):
        return hash((self.vendor_name, self.sbom_url or ""))


@dataclass(frozen=True)
class QuestionnaireResponse:
    """Single questionnaire response record."""
    vendor_id: str
    question_id: str
    answer_type: str  # text, yes/no, multiple-choice, file_upload
    answer_value: Optional[Any] = None
    confidence_score: float = 1.0  # 0-1, how certain we are
    verified_by: Optional[str] = None
    verification_date: Optional[datetime] = None
    
    @property
    def is_verified(self) -> bool:
        return self.verified_by is not None and self.verification_date is not None


@dataclass(frozen=True)
class VendorRiskProfile:
    """Aggregated risk profile for a vendor."""
    vendor_id: str
    base_score: float = 50.0  # Starting score, 0-100
    questionnaire_score: Optional[float] = None
    sbom_coverage: float = 0.0  # 0.0 to 1.0
    historical_incidents: int = 0
    last_assessment_date: datetime = field(default_factory=datetime.now)
    
    @property
    def risk_level(self) -> RiskLevel:
        """Determine risk level from base score."""
        if self.base_score >= 90:
            return RiskLevel.CRITICAL
        elif self.base_score >= 75:
            return RiskLevel.HIGH
        elif self.base_score >= 60:
            return RiskLevel.MEDIUM
        elif self.base_score >= 25:
            return RiskLevel.LOW
        else:
            return RiskLevel.MINIMAL
    
    def adjust(self, delta: float) -> None:
        """Adjust score by delta (can be negative)."""
        new_score = max(0.0, min(100.0, self.base_score + delta))
        _logger.debug("Adjusted %s from %.2f to %.2f", 
                     self.vendor_id, self.base_score, new_score)
        self.base_score = new_score


@dataclass(frozen=True)
class Vendor:
    """Complete vendor record."""
    profile: VendorProfile
    contact: Optional[Contact] = None
    risk_profile: VendorRiskProfile = field(default_factory=VendorRiskProfile)
    sbom_references: List[SBOMReference] = field(default_factory=list)
    questionnaire_responses: List[QuestionnaireResponse] = field(default_factory=list)
    audit_log: List[Dict[str, Any]] = field(default_factory=list)
    
    @property
    def effective_risk_score(self) -> float:
        """Calculate current risk score considering all factors."""
        if self.risk_profile.questionnaire_score is not None:
            # Blend questionnaire with base risk
            return (self.risk_profile.base_score * 0.4 + 
                    self.risk_profile.questionnaire_score * 0.6)
        return self.risk_profile.base_score
    
    def add_response(self, response: QuestionnaireResponse) -> None:
        """Add a new questionnaire response."""
        self.questionnaire_responses.append(response)
        self._audit_log_action("response_added", {
            "question_id": response.question_id,
            "timestamp": datetime.now().isoformat()
        })
    
    def _audit_log_action(self, action: str, details: Dict[str, Any]) -> None:
        """Log an audit event."""
        self.audit_log.append({
            "action": action,
            "details": details,
            "timestamp": datetime.now().isoformat(),
            "vendor_id": self.profile.unique_id
        })


class VendorRegistry(Generic[T]):
    """Thread-safe vendor registry with SBOM cross-referencing."""
    
    def __init__(self, data_dir: Optional[str] = None):
        """Initialize the registry.
        
        Args:
            data_dir: Directory for persisting registry state. Defaults to 
                     ~/.vendorvet/data if not specified.
        """
        self._vendors: Dict[str, Vendor] = {}
        self._data_dir = data_dir or os.path.expanduser(
            "~/.vendorvet/data"
        )
        os.makedirs(self._data_dir, exist_ok=True)
        
        # Load existing state if available
        self._load_from_disk()
    
    def _get_state_file_path(self) -> str:
        """Get path to the main state file."""
        return os.path.join(self._data_dir, "registry.json")
    
    def _save_to_disk(self) -> None:
        """Persist current state to disk."""
        try:
            with open(self._get_state_file_path(), 'w') as f:
                # Convert dataclasses to dicts for JSON serialization
                state = {
                    "vendors": [
                        self._vendor_to_dict(v) 
                        for v in sorted(
                            self._vendors.values(), 
                            key=lambda x: x.profile.unique_id
                        )
                    ],
                    "metadata": {
                        "version": "1.0",
                        "last_saved": datetime.now().isoformat()
                    }
                }
                json.dump(state, f, indent=2, default=str)
            _logger.info("Registry state saved to %s", self._get_state_file_path())
        except (IOError, OSError) as e:
            _logger.error("Failed to save registry: %s", e)
    
    def _vendor_to_dict(self, vendor: Vendor) -> Dict[str, Any]:
        """Convert a Vendor instance to a serializable dict."""
        return {
            "profile": asdict(vendor.profile),
            "contact": asdict(vendor.contact) if vendor.contact else None,
            "risk_profile": {
                "base_score": vendor.risk_profile.base_score,
                "questionnaire_score": vendor.risk_profile.questionnaire_score,
                "sbom_coverage": vendor.risk_profile.sbom_coverage,
                "historical_incidents": vendor.risk_profile.historical_incidents,
            },
            "sbom_references": [asdict(r) for r in vendor.sbom_references],
            "audit_log": vendor.audit_log[-10:],  # Keep last 10 entries only
        }
    
    def _load_from_disk(self) -> None:
        """Load existing registry state from disk."""
        state_file = self._get_state_file_path()
        if os.path.exists(state_file):
            try:
                with open(state_file, 'r') as f:
                    data = json.load(f)
                    
                    for v_data in data.get("vendors", []):
                        vendor = self._dict_to_vendor(v_data)
                        self._vendors[vendor.profile.unique_id] = vendor
                    
                    # Restore metadata if present
                    if "metadata" in data:
                        _logger.info(
                            "Loaded %d vendors from persisted state", 
                            len(self._vendors)
                        )
            except (json.JSONDecodeError, IOError) as e:
                _logger.warning("Corrupted state file, starting fresh: %s", e)
    
    def _dict_to_vendor(self, data: Dict[str, Any]) -> Vendor:
        """Convert a dict back to a Vendor instance."""
        profile = VendorProfile(**data["profile"])
        
        vendor = Vendor(
            profile=profile,
            contact=Contact(**data["contact"]) if data.get("contact") else None,
            risk_profile=VendorRiskProfile(
                vendor_id=profile.unique_id,
                base_score=data["risk_profile"]["base_score"],
                questionnaire_score=data["risk_profile"].get("questionnaire_score"),
                sbom_coverage=data["risk_profile"].get("sbom_coverage", 0.0),
                historical_incidents=data["risk_profile"].get(
                    "historical_incidents", 0
                ),
            )
        )
        
        # Restore SBOM references and audit log
        for r_data in data.get("sbom_references", []):
            vendor.sbom_references.append(SBOMReference(**r_data))
        
        vendor.audit_log = data.get("audit_log", [])
        return vendor
    
    def get_vendor(self, identifier: str) -> Optional[Vendor]:
        """Get a vendor by unique ID or name.
        
        Args:
            identifier: Vendor's unique_id, name, or partial match
            
        Returns:
            The matching Vendor instance, or None if not found
        """
        # Direct lookup by unique_id first
        if identifier in self._vendors:
            return self._vendors[identifier]
        
        # Fuzzy search by name
        for vendor in self._vendors.values():
            if (identifier.lower() in vendor.profile.name.lower() or
                identifier == vendor.profile.unique_id):
                return vendor
        
        _logger.debug("No vendor found matching '%s'", identifier)
        return None
    
    def add_vendor(self, vendor: Vendor) -> bool:
        """Add a new vendor to the registry.
        
        Args:
            vendor: The Vendor instance to add
            
        Returns:
            True if added successfully, False if duplicate exists
        """
        existing = self.get_vendor(vendor.profile.unique_id)
        if existing:
            _logger.warning("Vendor %s already exists", vendor.profile.name)
            return False
        
        self._vendors[vendor.profile.unique_id] = vendor
        self._save_to_disk()
        _logger.info("Added new vendor: %s", vendor.profile.name)
        return True
    
    def update_vendor(self, vendor: Vendor) -> bool:
        """Update an existing vendor's data.
        
        Args:
            vendor: The updated Vendor instance
            
        Returns:
            True if updated, False if not found
        """
        identifier = vendor.profile.unique_id
        if identifier in self._vendors:
            old_vendor = self._vendors[identifier]
            
            # Preserve audit history from the existing record
            new_audit_log = old_vendor.audit_log + [
                {
                    "action": "update",
                    "details": {"field": "full_record"},
                    "timestamp": datetime.now().isoformat(),
                    "vendor_id": identifier,
                }
            ]
            
            self._vendors[identifier] = vendor
            vendor.audit_log = new_audit_log
            
            # Only save if something actually changed
            if old_vendor.effective_risk_score != vendor.effective_risk_score:
                _logger.info("Updated risk score for %s", vendor.profile.name)
            
            self._save_to_disk()
            return True
        
        _logger.warning("Update failed - vendor %s not found", identifier)
        return False
    
    def remove_vendor(self, identifier: str) -> bool:
        """Remove a vendor from the registry.
        
        Args:
            identifier: Vendor's unique_id or name
            
        Returns:
            True if removed, False if not found
        """
        vendor = self.get_vendor(identifier)
        if not vendor:
            return False
        
        del self._vendors[vendor.profile.unique_id]
        self._save_to_disk()
        _logger.info("Removed vendor: %s", vendor.profile.name)
        return True
    
    def search(
        self, 
        name: Optional[str] = None,
        industry: Optional[str] = None,
        min_risk_score: float = 0.0,
        max_risk_score: float = 100.0,
        compliance_status: Optional[ComplianceStatus] = None,
    ) -> List[Vendor]:
        """Search vendors with multiple filters.
        
        Args:
            name: Partial or exact name match (case-insensitive)
            industry: Filter by industry code/name
            min_risk_score: Minimum effective risk score
            max_risk_score: Maximum effective risk score  
            compliance_status: Optional compliance filter
            
        Returns:
            List of matching Vendor instances
        """
        results = list(self._vendors.values())
        
        if name:
            search_term = name.lower()
            results = [v for v in results 
                      if search_term in v.profile.name.lower()]
        
        if industry:
            results = [v for v in results 
                      if industry.lower() in (v.profile.industry or "").lower()]
        
        if min_risk_score > 0 or max_risk_score < 100:
            results = [
                v for v in results 
                if min_risk_score <= v.effective_risk_score <= max_risk_score
            ]
        
        return results
    
    def get_sbom_coverage(self, vendor_id: str) -> float:
        """Calculate SBOM coverage for a specific vendor.
        
        Args:
            vendor_id: The unique identifier of the vendor
            
        Returns:
            Coverage ratio (0.0 to 1.0), or 0.0 if no products found
        """
        vendor = self.get_vendor(vendor_id)
        if not vendor:
            return 0.0
        
        # Calculate coverage based on known software products vs expected
        total_products = len(vendor.sbom_references)
        if total_products == 0:
            return 0.0
        
        # Assume we expect at least 3 major product lines for enterprise vendors
        expected_products = max(3, vendor.profile.size_category in ("L", "Enterprise") and 5 or 2)
        
        coverage = min(1.0, total_products / expected_products)
        vendor.risk_profile.sbom_coverage = coverage
        
        return coverage
    
    def cross_reference_sboms(self, 
                             software_name: str,
                             version: Optional[str] = None) -> List[Vendor]:
        """Find all vendors supplying a specific software product.
        
        Args:
            software_name: The software/product name to search for
            version: Optional exact version match
            
        Returns:
            List of Vendor instances that supply the matching software
        """
        results = []
        search_term = software_name.lower()
        
        for vendor in self._vendors.values():
            for sb

File: polyglot/go/vendor_registry.go
package main

import (
	"encoding/json"
	"fmt"
	"os"
	"sort"
	"strings"
	"sync"
	"time"
)

// Vendor represents a third-party vendor in the registry
type Vendor struct {
	ID          string        `json:"id"`
	Name        string        `json:"name"`
	Categories  []string      `json:"categories,omitempty"`
	RiskScore   float64       `json:"risk_score"`
	LastUpdated time.Time     `json:"last_updated"`
	Metadata    map[string]any `json:"metadata,omitempty"`
}

// SBOMEntry represents a software component from an SBOM linked to a vendor
type SBOMEntry struct {
	VendorID      string   `json:"vendor_id"`
	ComponentName string   `json:"component_name"`
	Version       string   `json:"version"`
	License       string   `json:"license,omitempty"`
	PackageMgr    string   `json:"package_manager,omitempty"`
}

// Registry holds the vendor registry state and operations
type Registry struct {
	mu         sync.RWMutex
	vendors    map[string]*Vendor
	sbomCache  map[string][]SBOMEntry
	thresholds map[string]float64 // "high", "medium", "low" risk thresholds
}

// NewRegistry creates a new vendor registry with default thresholds
func NewRegistry() *Registry {
	r := &Registry{
		vendors:    make(map[string]*Vendor),
		sbomCache:  make(map[string][]SBOMEntry),
		thresholds: map[string]float64{
			"high":   8.0,
			"medium": 5.0,
			"low":    3.0,
		},
	}
	return r
}

// RiskLevel returns the risk level string for a given score
func (r *Registry) RiskLevel(score float64) string {
	if score >= r.thresholds["high"] {
		return "HIGH"
	} else if score >= r.thresholds["medium"] {
		return "MEDIUM"
	}
	return "LOW"
}

// AddVendor creates a new vendor entry with validation
func (r *Registry) AddVendor(v Vendor) error {
	r.mu.Lock()
	defer r.mu.Unlock()

	if v.ID == "" {
		v.ID = fmt.Sprintf("vendor-%d", time.Now().UnixNano())
	}

	if v.Name == "" {
		return fmt.Errorf("vendor name required")
	}

	if _, exists := r.vendors[v.ID]; exists {
		return fmt.Errorf("vendor ID %s already exists", v.ID)
	}

	v.LastUpdated = time.Now()
	r.vendors[v.ID] = &v

	return nil
}

// GetVendor retrieves a vendor by ID with thread safety
func (r *Registry) GetVendor(id string, includeSBOM bool) (*Vendor, error) {
	r.mu.RLock()
	defer r.mu.RUnlock()

	v, exists := r.vendors[id]
	if !exists {
		return nil, fmt.Errorf("vendor %s not found", id)
	}

	if includeSBOM && v != nil {
		sbomEntries := make([]SBOMEntry, 0, len(v.SBOMEntries))
		for _, e := range v.SBOMEntries {
			if e.VendorID == id {
				sbomEntries = append(sbomEntries, e)
			}
		}
		v.SBOMEntries = sbomEntries
	}

	return v, nil
}

// SearchVendors performs a case-insensitive search across multiple fields
func (r *Registry) SearchVendors(query string) ([]Vendor, error) {
	r.mu.RLock()
	defer r.mu.RUnlock()

	if query == "" {
		vendors := make([]Vendor, 0, len(r.vendors))
		for _, v := range r.vendors {
			vendors = append(vendors, *v)
		}
		return vendors, nil
	}

	queryLower := strings.ToLower(query)
	results := make([]Vendor, 0)

	for _, v := range r.vendors {
		if strings.Contains(strings.ToLower(v.Name), queryLower) ||
			strings.Contains(strings.Join(v.Categories, " "), queryLower) {
			results = append(results, *v)
		}
	}

	sort.Slice(results, func(i, j int) bool {
		return results[i].RiskScore > results[j].RiskScore
	})

	return results, nil
}

// CrossRefSBOM parses an SBOM and finds all matching vendors in the registry
func (r *Registry) CrossRefSBOM(sbomData []byte) ([]SBOMEntry, error) {
	var entries []SBOMEntry

	err := json.Unmarshal(sbomData, &entries)
	if err != nil {
		return nil, fmt.Errorf("failed to parse SBOM: %w", err)
	}

	matched := make([]SBOMEntry, 0)
	for _, e := range entries {
		vendor, exists := r.GetVendor(e.VendorID, false)
		if exists && vendor != nil {
			e.RiskLevel = r.RiskLevel(vendor.RiskScore)
			matched = append(matched, e)
		} else if !exists {
			e.Status = "UNKNOWN_VENDOR"
			matched = append(matched, e)
		}
	}

	return matched, nil
}

// UpdateRiskScore modifies a vendor's risk score with validation
func (r *Registry) UpdateRiskScore(id string, newScore float64) error {
	r.mu.Lock()
	defer r.mu.Unlock()

	v, exists := r.vendors[id]
	if !exists {
		return fmt.Errorf("vendor %s not found", id)
	}

	oldScore := v.RiskScore
	v.RiskScore = newScore
	v.LastUpdated = time.Now()

	return nil
}

// GetRiskSummary returns a summary of all vendors grouped by risk level
func (r *Registry) GetRiskSummary() map[string][]*Vendor {
	r.mu.RLock()
	defer r.mu.RUnlock()

	summary := make(map[string][]*Vendor, 3)
	for _, v := range r.vendors {
		level := r.RiskLevel(v.RiskScore)
		summary[level] = append(summary[level], v)
	}

	return summary
}

// ExportJSON marshals the entire registry to JSON bytes
func (r *Registry) ExportJSON() ([]byte, error) {
	r.mu.RLock()
	defer r.mu.RUnlock()

	type Export struct {
		Vendors    []Vendor      `json:"vendors"`
		SBOMCache  map[string][]SBOMEntry `json:"sbom_cache,omitempty"`
		Thresholds map[string]float64 `json:"thresholds"`
	}

	export := Export{
		Vendors:     make([]Vendor, 0, len(r.vendors)),
		Thresholds:  r.thresholds,
		SBOMCache:   r.sbomCache,
	}

	for _, v := range r.vendors {
		export.Vendors = append(export.Vendors, *v)
	}

	return json.MarshalIndent(&export, "", "  ")
}

// ImportJSON loads vendors from JSON bytes
func (r *Registry) ImportJSON(data []byte) error {
	var export Export
	err := json.Unmarshal(data, &export)
	if err != nil {
		return fmt.Errorf("failed to parse registry JSON: %w", err)
	}

	r.mu.Lock()
	defer r.mu.Unlock()

	for _, v := range export.Vendors {
		v.ID = fmt.Sprintf("%s-%d", v.ID, time.Now().UnixNano())
		r.vendors[v.ID] = &v
	}

	return nil
}

// VendorService provides a service layer for vendor operations
type VendorService struct {
	registry *Registry
}

// NewVendorService creates a new service instance
func NewVendorService(registry *Registry) *VendorService {
	return &VendorService{registry: registry}
}

// CreateVendor wraps AddVendor with error handling and logging
func (s *VendorService) CreateVendor(name string, categories []string, riskScore float64) (*Vendor, error) {
	v := Vendor{
		Name:       name,
		Categories: categories,
		RiskScore:  riskScore,
	}

	if err := s.registry.AddVendor(v); err != nil {
		return nil, fmt.Errorf("failed to create vendor %s: %w", name, err)
	}

	return &v, nil
}

// FindVendorsByCategory returns all vendors in a specific category
func (s *VendorService) FindVendorsByCategory(category string) ([]*Vendor, error) {
	results := make([]*Vendor, 0)
	s.registry.mu.RLock()
	defer s.registry.mu.RUnlock()

	for _, v := range s.registry.vendors {
		for _, c := range v.Categories {
			if strings.EqualFold(c, category) {
				results = append(results, v)
				break
			}
		}
	}

	return results, nil
}

// CalculateAggregateRisk computes the average risk score across all vendors
func (s *VendorService) CalculateAggregateRisk() float64 {
	s.registry.mu.RLock()
	defer s.registry.mu.RUnlock()

	if len(s.registry.vendors) == 0 {
		return 0.0
	}

	var total float64
	for _, v := range s.registry.vendors {
		total += v.RiskScore
	}

	return total / float64(len(s.registry.vendors))
}

// Main demo function showcasing all capabilities
func main() {
	fmt.Println("=== VendorVet Registry Demo ===\n")

	reg := NewRegistry()
	service := NewVendorService(reg)

	// 1. Create vendors
	fmt.Println("--- Creating Vendors ---")
	vendor1, _ := service.CreateVendor(
		"Acme Corp",
		[]string{"cloud", "infrastructure"},
		7.5,
	)
	vendor2, _ := service.CreateVendor(
		"TechSolutions Inc",
		[]string{"software", "saas"},
		3.2,
	)

	fmt.Printf("Created: %+v\n", vendor1)
	fmt.Printf("Risk Level: %s\n\n", reg.RiskLevel(vendor1.RiskScore))

	// 2. Search and retrieve
	fmt.Println("--- Searching Vendors ---")
	results, _ := reg.SearchVendors("cloud")
	for _, v := range results {
		fmt.Printf("Found: %s (Risk: %.1f)\n", v.Name, v.RiskScore)
	}

	vendor, _ := reg.GetVendor(vendor1.ID, true)
	fmt.Printf("\nFull Vendor Record:\n%+v\n\n", vendor)

	// 3. Risk summary
	fmt.Println("--- Risk Summary ---")
	summary := reg.GetRiskSummary()
	for level, vendors := range summary {
		fmt.Printf("%s: %d vendors (avg risk: %.2f)\n", level, len(vendors), calculateAvgRisk(vendors))
	}

	// 4. Export/Import simulation
	fmt.Println("\n--- Export/Import Demo ---")
	exported, _ := reg.ExportJSON()
	fmt.Printf("Exported size: %d bytes\n", len(exported))

	// 5. Aggregate calculation
	fmt.Printf("\nAggregate Risk Score: %.2f\n", service.CalculateAggregateRisk())

	// 6. Update risk score
	fmt.Println("\n--- Updating Risk Scores ---")
	reg.UpdateRiskScore(vendor1.ID, 9.0)
	vendor3, _ := reg.GetVendor(vendor1.ID, false)
	fmt.Printf("Updated risk: %.1f (Level: %s)\n", vendor3.RiskScore, reg.RiskLevel(vendor3.RiskScore))

	// 7. Final state
	fmt.Println("\n--- Final Registry State ---")
	exportedFinal, _ := reg.ExportJSON()
	var finalExport Export
	json.Unmarshal(exportedFinal, &finalExport)
	for _, v := range finalExport.Vendors {
		fmt.Printf("ID: %s | Name: %-20s | Risk: %.1f | Level: %s\n",
			v.ID, v.Name, v.RiskScore, reg.RiskLevel(v.RiskScore))
	}

	// 8. Error handling demo
	fmt.Println("\n--- Error Handling Demo ---")
	_, err := reg.GetVendor("nonexistent-id-12345", false)
	if err != nil {
		fmt.Printf("Expected error: %v\n", err)
	}

	_, err = service.CreateVendor("", []string{}, 0.0)
	if err != nil {
		fmt.Printf("Validation error: %v\n", err)
	}
}

// Helper function for risk calculation
func calculateAvgRisk(vendors []*Vendor) float64 {
	if len(vendors) == 0 {
		return 0.0
	}
	var total float64
	for _, v := range vendors {
		total += v.RiskScore
	}
	return total / float64(len(vendors))
}

// Export type alias for main function
type Export struct {
	Vendors    []Vendor      `json:"vendors"`
	SBOMCache  map[string][]SBOMEntry `json:"sbom_cache,omitempty"`
	Thresholds map[string]float64 `json:"thresholds"`
}